pkg/block/node.go

```diff
@@ -0,0 +1,87 @@
+// Copyright (c) The Thanos Authors.
+// Licensed under the Apache License 2.0.
+
+package block
+
+import (
+	"github.com/oklog/ulid"
+	"github.com/thanos-io/thanos/pkg/block/metadata"
+)
+
+type compareNodes func(parentMeta, childMeta *metadata.Meta) bool
+
+// Node type implements a node of a tree.
+type Node struct {
+	ID       ulid.ULID
+	Value    *metadata.Meta
+	Children []*Node
+}
+
+// NewNode creates a new node.
+func NewNode(id ulid.ULID, value *metadata.Meta) *Node {
+	return &Node{
+		ID:       id,
+		Children: []*Node{},
+		Value:    value,
+	}
+}
+
+// AddMeta Inserts meta as a node such that the Sources of parent node is superset of all child vertices.
+func (n *Node) AddMeta(id ulid.ULID, metas map[ulid.ULID]*metadata.Meta, areEqual, isChildNode compareNodes) bool {
+	return n.add(id, n, metas, areEqual, isChildNode)
+}
+
+// GetNonRootIDs returns list of ids which are not on root level.
+func (n *Node) GetNonRootIDs() []ulid.ULID {
+	var ulids []ulid.ULID
+	for _, node := range n.Children {
+		ulids = append(ulids, n.iterate(node)...)
+		ulids = remove(ulids, node.ID)
+	}
+	return ulids
+}
+
+func (n *Node) iterate(node *Node) []ulid.ULID {
+	var ulids []ulid.ULID
+	ulids = append(ulids, node.ID)
+	for _, childNode := range node.Children {
+		ulids = append(ulids, n.iterate(childNode)...)
+	}
+	return ulids
+}
+
+func (n *Node) add(id ulid.ULID, startNode *Node, metas map[ulid.ULID]*metadata.Meta, areEqual, isChildNode compareNodes) bool {
```

Comment:
Ok, I said like 4 times to move this from node and it's still there and you did not explain why you are against this (: 

Maybe I can send PR to your PR to show you how it can be done and show adventages? (:

Reply:
The main problem is that this is just a `node` 

It's not  `hierarchicalBlocksNode` or anything like this. Method is also just `add` which suggests nothing is done expect just adding a node to the tree ... somewhere. This is might be not clear for readers of this code so we need to improve this.

Second argument why I believe it would nice to move `add` to separate function is extensibility. What if we would like to use our tree with meta.jsons data struct for other algorithms? We can't because `add` has certain logic baked in.